applications/nrf_desktop/src/modules/hid_state.c

```diff
@@ -814,32 +828,31 @@ static bool update_report(struct report_data *rd)
 					      event->item.usage_id,
 					      event->item.value);
 
-		rd->update_needed = rd->update_needed || update_needed;
+		rd->linked_rs->update_needed = rd->linked_rs->update_needed || update_needed;
 
 
 		k_free(event);
 
 		/* If no item was changed, try next event. */
 	}
 
-	if (rd->update_needed) {
-		update_needed = true;
-	}
-
-	return update_needed;
+	return rd->linked_rs->update_needed;
 }
 
-static bool report_send(struct report_data *rd, bool check_state, bool send_always)
+static bool report_send(struct report_data *rd,
+			struct report_state *rs,
+			bool check_state,
+			bool send_always)
 {
 	bool report_sent = false;
-
-	if (!rd->linked_rs) {
+	if (!rs && rd) {
+		rs = rd->linked_rs;
+	}
```

Comment:
we could add here following to emphasize at call time that we either go from rs or rd point
add after line 850
```
if (!rd && rs) {
   rd = rs->linked_rd;
}
if (!rs || !rd) {
   LOG_WRN("No linked report sink");
   return report_sent;
}
```